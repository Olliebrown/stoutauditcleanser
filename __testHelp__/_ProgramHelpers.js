import fs from 'fs'
import path from 'path'

// Project specific imports
import { QUERIES } from '../src/domTraversal/queriesAndRegex.js'
import Program from '../src/Objects/Program.js'

export function InitProgramTest (htmlFileName) {
  // Read in the HTML page
  const baseHTML = fs.readFileSync(
    path.join('__tests__', 'data', 'ProgramTables', htmlFileName),
    'utf8'
  )

  // Load page from imported HTML
  document.documentElement.innerHTML = baseHTML.toString()

  // Extract program headers and make sure there's at least one
  const headerNodes = document.querySelectorAll(QUERIES.programHeader)
  if (headerNodes?.length < 1) {
    throw new Error('Unable to find at least one program')
  }

  // Create program object
  return new Program(headerNodes[0])
}
