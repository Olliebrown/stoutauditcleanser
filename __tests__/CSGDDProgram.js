// Necessary when using ECMA modules
import { jest } from '@jest/globals'

// Help load the program HTML page
import { InitProgramTest } from '../__testHelp__/_ProgramHelpers.js'

// Disable jest mocking for fs module
jest.dontMock('fs')

describe('BS in CS: Game Des & Dev Program Node', () => {
  // Inject clean version of HTML page and rebuild program before tests
  let program = null
  beforeAll(() => {
    // Load page from imported HTML
    program = InitProgramTest('CSGDDProgram.html')
  })

  // Clear out the HTML and program
  afterAll(() => {
    jest.resetModules()
    program = null
  })

  // Basic page structure tests
  test('should have a specific page title and specific program name', () => {
    // Check title
    const title = document.querySelector('title')
    expect(title).toBeTruthy()
    expect(title.textContent).toBe('BS Computer Science Game Design & Dev Program')

    // Check program name
    expect(program).toBeTruthy()
    expect(program.getName()).toMatch(/^BS In CS: GDD\s+\d\d\d\d$/)
  })
})
