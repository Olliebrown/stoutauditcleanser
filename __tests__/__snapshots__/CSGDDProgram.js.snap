// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BS in CS: Game Des & Dev Program Node should have expected snapshot 1`] = `
{
  "internalId": "RG-0783",
  "isGenEd": false,
  "key": "bs_in_cs__gdd__2019",
  "name": "BS In CS: GDD  2019",
  "satisfied": "INCOMPLETE",
  "subNodes": [
    {
      "description": "Racial and Ethnic StudiesStudents must take 6 credits, a minimum of 3 credits must be from RES-A",
      "index": 1,
      "internalId": "RQ-10360",
      "isGenEd": true,
      "key": "res_requirement",
      "name": "RES Requirement",
      "nextIndex": 11,
      "satisfied": "INCOMPLETE",
      "subNodes": [
        {
          "description": "Option 1: Take two RES-A Course",
          "internalId": "LN-10",
          "isGenEd": true,
          "isValid": true,
          "key": "option_1_take_two_res_a_course",
          "name": "Option 1:take Two RES-a Course",
          "satisfied": "INCOMPLETE",
          "subNodes": [],
          "units": {},
        },
        {
          "description": "Option 2: Take one RES-A",
          "internalId": "LN-020",
          "isGenEd": true,
          "isValid": true,
          "key": "option_2__take_one_res_a",
          "name": "Option 2: Take One RES-a",
          "satisfied": "INCOMPLETE",
          "subNodes": [],
          "units": {},
        },
        {
          "description": "AND one RES-B course",
          "internalId": "LN-030",
          "isGenEd": true,
          "isValid": true,
          "key": "__one_res_b_course",
          "name": "& One RES-b Course",
          "satisfied": "INCOMPLETE",
          "subNodes": [],
          "units": {},
        },
        {
          "description": "Option 3: One Transfer Ethnic StudyNOTE:  This requirement is only meant for transfer students who have met the requirement at their previous institution. ",
          "internalId": "LN-040",
          "isGenEd": false,
          "isValid": true,
          "key": "option_3__one_transfer_ethnic_study",
          "name": "Option 3: One Transfer Ethnic Study",
          "satisfied": "INCOMPLETE",
          "subNodes": [],
          "units": {},
        },
      ],
      "units": {
        "courses": {
          "current": 0,
          "required": 1,
        },
        "credits": {
          "current": 0,
          "required": 6,
        },
      },
    },
    {
      "description": "Global Perspectives",
      "index": 11,
      "internalId": "RQ-025",
      "isGenEd": true,
      "key": "glp_requirement",
      "name": "GLP Requirement",
      "nextIndex": 15,
      "satisfied": "COMPLETE",
      "subNodes": [
        {
          "description": "Global Perspective Courses  - 6 Credits",
          "internalId": "LN-010",
          "isGenEd": false,
          "isValid": true,
          "key": "global_perspective_courses",
          "name": "Global Perspective Courses",
          "satisfied": "COMPLETE",
          "subNodes": [],
          "units": {},
        },
      ],
      "units": {
        "credits": {
          "current": 12,
          "required": 6,
        },
      },
    },
    {
      "description": "General Education: Communication Skills",
      "index": 15,
      "internalId": "RQ-020",
      "isGenEd": true,
      "key": "gened__communication_skills",
      "name": "GenEd: Communication Skills",
      "nextIndex": 23,
      "satisfied": "COMPLETE",
      "subNodes": [
        {
          "description": "Take ENGL-101, ENGL 101HON or ENGL-111HON",
          "internalId": "LN-010",
          "isGenEd": false,
          "isValid": true,
          "key": "freshman_composition_1___engl_101__or_engl_101hon_or_engl_111hon",
          "name": "Freshman Composition 1 - ENGL-101, Or ENGL 101hon Or ENGL-111hon",
          "satisfied": "COMPLETE",
          "subNodes": [],
          "units": {},
        },
        {
          "description": "Take ENGL-102, ENGL-102HON or ENGL-113HON",
          "internalId": "LN-020",
          "isGenEd": false,
          "isValid": true,
          "key": "freshman_composition_2___engl_102__engl_102hon_or_engl_113hon",
          "name": "Freshman Composition 2 - ENGL-102, ENGL-102hon Or ENGL-113hon",
          "satisfied": "COMPLETE",
          "subNodes": [],
          "units": {},
        },
        {
          "description": "Take COMST 100",
          "internalId": "LN-030",
          "isGenEd": false,
          "isValid": true,
          "key": "fundamentals_of_speech___comst_100",
          "name": "Fundamentals Of Speech - COMST 100",
          "satisfied": "COMPLETE",
          "subNodes": [],
          "units": {},
        },
      ],
      "units": {},
    },
    {
      "description": "General Education-Analytical Reasoning & Natural ScienceComputer Science- Game Design & Development Requirements",
      "index": 23,
      "internalId": "RQ-10431",
      "isGenEd": true,
      "key": "gened__arns",
      "name": "GenEd: ARNS",
      "nextIndex": 33,
      "satisfied": "IN_PROGRESS",
      "subNodes": [
        {
          "description": "Take CS 144Grade of C or better required",
          "internalId": "UNKNOWN_ID",
          "isGenEd": false,
          "isValid": true,
          "key": "take_cs_144",
          "name": "Take CS 144",
          "satisfied": "COMPLETE",
          "subNodes": [],
          "units": {},
        },
        {
          "description": "Take MATH 151 Or MATH 153 OR MATH 156Grade of C or better required",
          "internalId": "LN-015",
          "isGenEd": false,
          "isValid": true,
          "key": "take_math_151_or_math_153_or_math_156",
          "name": "Take MATH 151 Or MATH 153 Or MATH 156",
          "satisfied": "COMPLETE",
          "subNodes": [],
          "units": {},
        },
        {
          "description": "PHYS 281",
          "internalId": "LN-020",
          "isGenEd": false,
          "isValid": true,
          "key": "take_phys_281",
          "name": "Take PHYS 281",
          "satisfied": "INCOMPLETE",
          "subNodes": [],
          "units": {},
        },
      ],
      "units": {
        "credits": {
          "current": 8,
          "required": 13,
        },
      },
    },
    {
      "description": "General Education: Humanities and the Arts- Select from two or more areas Including: Art History/Music Appreciation, Creative/Performing Arts, History, Literature, Philosophy, & Foreign Language & Culture  *Effective 01/01/2021, PHOTO 130 NO LONGER classified as GE HUM/ART & does not satisfy this area.   **Effective 08/15/2021, ART 100 & ART 101 NO LONGER classified as GE HUM/ART & do not satisfy this area.",
      "index": 33,
      "internalId": "RQ-019",
      "isGenEd": true,
      "key": "gened__humart",
      "name": "GenEd: HumArt",
      "nextIndex": 41,
      "satisfied": "COMPLETE",
      "subNodes": [
        {
          "description": "Art History/Music Appreciation Option ",
          "internalId": "LN-005",
          "isGenEd": false,
          "isValid": true,
          "key": "art_history_music_appreciation",
          "name": "Art History/music Appreciation",
          "satisfied": "COMPLETE",
          "subNodes": [],
          "units": {},
        },
        {
          "description": "Foreign Language & Culture ",
          "internalId": "LN-050",
          "isGenEd": false,
          "isValid": true,
          "key": "foreign_language___culture",
          "name": "Foreign Language & Culture",
          "satisfied": "COMPLETE",
          "subNodes": [],
          "units": {},
        },
      ],
      "units": {
        "credits": {
          "current": 8,
          "required": 6,
        },
      },
    },
    {
      "description": "General Education: Social and Behavioral Sciences - Courses MUST be from two or more Areas Including: Anthropology, Economics, Geography, Political Science, Psychology and Sociology",
      "index": 41,
      "internalId": "RQ-0138",
      "isGenEd": true,
      "key": "gened__sbsci",
      "name": "GenEd: SBSci",
      "nextIndex": 47,
      "satisfied": "COMPLETE",
      "subNodes": [
        {
          "description": "Geography Option ",
          "internalId": "LN-050",
          "isGenEd": false,
          "isValid": true,
          "key": "geography_option",
          "name": "Geography Option",
          "satisfied": "COMPLETE",
          "subNodes": [],
          "units": {},
        },
      ],
      "units": {
        "credits": {
          "current": 3,
          "required": 6,
        },
      },
    },
    {
      "description": "General Education: Cross-Disciplinary Issues",
      "index": 47,
      "internalId": "RQ-10361",
      "isGenEd": true,
      "key": "gened__ci",
      "name": "GenEd: CI",
      "nextIndex": 53,
      "satisfied": "INCOMPLETE",
      "subNodes": [
        {
          "description": "Complete 3 credits from the approved CISS course list",
          "internalId": "LN-010",
          "isGenEd": false,
          "isValid": true,
          "key": "select_3_credits",
          "name": "Select 3 Credits",
          "satisfied": "INCOMPLETE",
          "subNodes": [],
          "units": {},
        },
      ],
      "units": {
        "credits": {
          "current": 0,
          "required": 3,
        },
      },
    },
    {
      "description": "General Education- Social Responsibility and Ethical ReasoningTake ICT 215",
      "index": 53,
      "internalId": "RQ-10432",
      "isGenEd": true,
      "key": "gened__srer",
      "name": "GenEd: SRER",
      "nextIndex": 57,
      "satisfied": "COMPLETE",
      "subNodes": [
        {
          "description": "ICT 215",
          "internalId": "UNKNOWN_ID",
          "isGenEd": false,
          "isValid": true,
          "key": "ict_215",
          "name": "ICT 215",
          "satisfied": "COMPLETE",
          "subNodes": [],
          "units": {},
        },
      ],
      "units": {
        "courses": {
          "current": 1,
          "required": 0,
        },
      },
    },
    {
      "description": "COMPUTER SCIENCE MAJOR STUDIES A grade of C or better is required for all MATH, CS, and GDD courses. ",
      "index": 57,
      "internalId": "RQ-0451",
      "isGenEd": false,
      "key": "maj__cs_major_studies",
      "name": "Maj: CS Major Studies",
      "nextIndex": 69,
      "satisfied": "IN_PROGRESS",
      "subNodes": [
        {
          "description": "Computer Science Courses",
          "internalId": "UNKNOWN_ID",
          "isGenEd": false,
          "isValid": true,
          "key": "cs_courses",
          "name": "CS Courses",
          "satisfied": "INCOMPLETE",
          "subNodes": [],
          "units": {},
        },
        {
          "description": "MATH 154 OR MATH 157",
          "internalId": "UNKNOWN_ID",
          "isGenEd": false,
          "isValid": true,
          "key": "math_154_or_math_157",
          "name": "MATH 154 Or MATH 157",
          "satisfied": "COMPLETE",
          "subNodes": [],
          "units": {},
        },
        {
          "description": "Mathematical CoursesTake MATH 270 & MATH  275",
          "internalId": "UNKNOWN_ID",
          "isGenEd": false,
          "isValid": true,
          "key": "mathematical_courses",
          "name": "Mathematical Courses",
          "satisfied": "INCOMPLETE",
          "subNodes": [],
          "units": {},
        },
        {
          "description": "Mathematics Selectives",
          "internalId": "UNKNOWN_ID",
          "isGenEd": false,
          "isValid": true,
          "key": "mathematics_selectives",
          "name": "Mathematics Selectives",
          "satisfied": "INCOMPLETE",
          "subNodes": [],
          "units": {},
        },
      ],
      "units": {
        "courses": {
          "current": 3,
          "required": 11,
        },
        "credits": {
          "current": 15,
          "required": 45,
        },
      },
    },
    {
      "description": "Student must complete a minimum of 6 Natural Science Credits",
      "index": 69,
      "internalId": "RQ-010760",
      "isGenEd": false,
      "key": "verify_completion_of_natural_science_credit_requirement",
      "name": "Verify Completion Of Natural Science Credit Requirement",
      "nextIndex": 73,
      "satisfied": "INCOMPLETE",
      "subNodes": [
        {
          "description": "Natural Science Courses",
          "internalId": "UNKNOWN_ID",
          "isGenEd": false,
          "isValid": true,
          "key": "natural_science_courses",
          "name": "Natural Science Courses",
          "satisfied": "INCOMPLETE",
          "subNodes": [],
          "units": {},
        },
      ],
      "units": {
        "credits": {
          "current": 0,
          "required": 6,
        },
      },
    },
    {
      "description": "GAME DESIGN AND DEVELOPMENT CONCENTRATION A grade of C or better is required for all MATH, MSCS, CS, and GDD courses. ",
      "index": 73,
      "internalId": "RQ-10509",
      "isGenEd": false,
      "key": "conc__gdd",
      "name": "CONC: GDD",
      "nextIndex": 81,
      "satisfied": "INCOMPLETE",
      "subNodes": [
        {
          "description": "Take the following 8 courses",
          "internalId": "UNKNOWN_ID",
          "isGenEd": false,
          "isValid": true,
          "key": "take_the_following_8_courses",
          "name": "Take The Following 8 Courses",
          "satisfied": "INCOMPLETE",
          "subNodes": [],
          "units": {},
        },
        {
          "description": "Game Design and Development Selectives",
          "internalId": "UNKNOWN_ID",
          "isGenEd": false,
          "isValid": true,
          "key": "gdd_selectives",
          "name": "GDD Selectives",
          "satisfied": "INCOMPLETE",
          "subNodes": [],
          "units": {},
        },
      ],
      "units": {
        "courses": {
          "current": 2,
          "required": 8,
        },
        "credits": {
          "current": 8,
          "required": 32,
        },
      },
    },
    {
      "description": "OTHER COURSES",
      "index": 81,
      "internalId": "RQ-0331",
      "isGenEd": false,
      "key": "other_courses",
      "name": "Other Courses",
      "nextIndex": 86,
      "satisfied": "COMPLETE",
      "subNodes": [
        {
          "description": "Courses Not Used Toward Degree",
          "internalId": "UNKNOWN_ID",
          "isGenEd": false,
          "isValid": true,
          "key": "courses_not_used_toward_degree",
          "name": "Courses Not Used Toward Degree",
          "satisfied": "COMPLETE",
          "subNodes": [],
          "units": {},
        },
        {
          "description": "Transfer Courses Not Used Toward Degree",
          "internalId": "UNKNOWN_ID",
          "isGenEd": false,
          "isValid": true,
          "key": "transfer_courses_not_used_toward_degree",
          "name": "Transfer Courses Not Used Toward Degree",
          "satisfied": "COMPLETE",
          "subNodes": [],
          "units": {},
        },
      ],
      "units": {},
    },
  ],
  "units": {
    "credits": {
      "current": 54,
      "required": 120,
    },
    "gpa": {
      "current": 3.819,
      "required": 2,
    },
  },
}
`;
